### ISSUE TITLE
can't found the window handle in handle view

### ISSUE THREAD

**Issue by oldn123** [x64dbg/x64dbg#2032]
hi, master
If the child window is the parent window of other processes, then after attach this process I can‘t found the window handle in handle view.

**Comment by pdcrrt** [comment]
Can you send some screens?

**Comment by mrexodia** [comment]
The code for enumerating is here, do you see the bug? https://github.com/x64dbg/x64dbg/blob/development/src/dbg/handles.cpp#L220

**Comment by Mattiwatti** [comment]
This is probably going to be borderline useless given my abysmal knowledge of Win32 GUI programming, but... of the two decent 'window spy' tools I know, one (your Spy++) uses DLL injection to set a hook, and the other (Process Hacker's Window Explorer) uses a [fairly convoluted method involving a shared section](https://github.com/processhacker/processhacker/blob/b4614ff45e4e9527645cdc949260a8ad9b9fdef8/plugins/WindowExplorer/hook.c) but which does seem to avoid straight out DLL injection. (NB: I linked to an older commit on purpose as the current PH codebase is completely broken on my machine and this file seems to have been removed as well...). The PH DLL runs in two separate 'modes' (client or server) depending on the process context and these communicate with each other to post/retrieve info on the window proc addresses and such.

So compared to this, the x64dbg approach suddenly seems rather tame. Not that I'm complaining :laughing: I'm not sure what the advantage of a hook DLL would be over simply enumerating windows out-of-process; maybe someone can tell me why these tools go to such lengths to enumerate other processs windows this way, no doubt there's a reason for it. But I think any solution requiring either DLL injection or section mapping would more likely fall into plugin territory than x64dbg core functionality.

OP, is it possible for you to post a small sample app (preferably not malware - if malware, please let me know in advance so I'll use my Windows 10 box) with some steps to reproduce? I'm not 100% clear on what order the process -> window creation is supposed to go, which process owns what window, and which process I should be debugging.

**Comment by oldn123** [comment]
I've just experimented with another set of applications, and the problem is the same as the topic. I've created two GUI processes, set the main window of the first process to the child window of the second process, and then look at the second process with x32dbg and spy ++ respectively, and compare the windows they listed. You can see the difference.

**Comment by mrexodia** [comment]
@oldn123 I believe there is a difference, but I was asking if you have any idea why this difference happens 🙂 Possibly x64dbg just simply doesn't enumerate child windows...

**Comment by Mattiwatti** [comment]
@mrexodia And that in turn was why I was wondering why all these 'specialized'/'good' (admittedly, N = 2) window inspectors that give accurate results seem to need a hook DLL :stuck_out_tongue: (not a rhetorical question, I have no idea.)

ping @reliasn

**Comment by oldn123** [comment]
@mrexodia yes, the EnumWindows just only enumerate all top-level windows , so it be missing .

**Comment by torusrxxx** [comment]
Child windows are followed and enumerated.
https://github.com/x64dbg/x64dbg/blob/da913f7cdc53f7d01293885f6706aaead14b2903/src/dbg/handles.cpp#L227

**Comment by torusrxxx** [comment]
A possible problem of the code is that when the debuggee creates a child window under a parent window which the debuggee doesn't own (ridiculous but not impossible). In this case that top level window will be dropped at first `EnumWindows` scan, and then this child window of debuggee is missed.
